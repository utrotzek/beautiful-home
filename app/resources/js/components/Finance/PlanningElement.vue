<template>
    <div
        ref="planningElement"
        class="planningElement form-group"
        @click="showOverlay"
    >
        <Progressbar
            ref="topProgress"
            color="#f8f9fa"
        />

        <div
            v-if="hasDate"
            class="date"
        >
            <span v-if="!localEditMode">{{ localPlanningItem.date | formatDate }}</span>
            <v-date-picker
                v-else
                v-model="date"
                :popover="{visibility: 'focus', placement: 'bottom', keepVisibleOnInput: false}"
                :available-dates="{ start: startDate, end: endDate }"
                :attributes="vCalendarAttributes"
            />
        </div>
        <div class="title">
            <div class="float-left">
                <span v-if="!localEditMode">{{ localPlanningItem.costCenter.title }}</span>
                <AutoCompleter
                    v-else
                    :items="costCenterData"
                    :enable-inline-creation="true"
                    :show-all-items-on-empty-query="true"
                    :preselected-value="localPlanningItem.costCenter.id"
                    search-key="title"
                    value-key="id"
                    placeholder="Kostenstelle wÃ¤hlen"
                    :class="errorClass($v.localPlanningItem.costCenter.id)"
                    @selected="costCenterUpdate"
                    @create="createCostCenter"
                />
            </div>
            <div
                class="float-right text-right"
                :class="negativePostiveClass"
            >
                <span v-if="!localEditMode">{{ localPlanningItem.totalAmount | toCurrency }}</span>
                <input
                    v-else
                    v-model="$v.localPlanningItem.totalAmount.$model"
                    type="text"
                    class="form-control text-right"
                    :class="[negativePostiveClass, errorClass($v.localPlanningItem.totalAmount)]"
                >
            </div>
        </div>
        <div class="clearfix"></div>
        <div class="description">
            <span v-if="!localEditMode" class="pre-formatted">{{ localPlanningItem.description }}</span>
            <textarea
                v-else
                v-model="localPlanningItem.description"
                rows="3"
                class="form-control"
            ></textarea>
        </div>

        <div
            v-if="displayOverlay"
            class="overlay"
        >
            <ButtonRow
                v-if="!localEditMode"
                show-delete
                show-edit
                show-close
                :show-connect="!isConnected"
                :disabled="$v.$anyError"
                @close="closeOverlay"
                @edit="editPlanning"
                @connect="connectPlanning"
                @deleteClick="deletePlanning"
            />

            <ButtonRow
                v-else
                show-save
                show-cancel
                :disabled="$v.$anyError"
                @close="closeOverlay"
                @save="saveEdit"
                @cancel="cancelEdit"
            />
        </div>
    </div>
</template>

<script>
import { required, integer } from "vuelidate/lib/validators";
import _ from "lodash";
import moment from "moment";
import ButtonRow from "../../../js/components/tools/ButtonRow";
import AutoCompleter from "../../../js/components/tools/Autocompleter";

const notZero = (value) => value > 0 || value < 0;
//99999 is the default entry for new planning elements
const notDefaultCostCenterId = (value) => value !== 99999;

export default {
    name: "PlanningElementVue",
    components: {
        ButtonRow,
        AutoCompleter
    },
    props: {
        year: {
            type: Number,
            required: true,
            default: 2020
        },
        month: {
            type: Number,
            required: true,
            default: 10
        },
        planningItem: {
            type: Object,
            required: true,
            default() {
                return {
                    id: 1,
                    costCenter: {
                        id: 100,
                        title: "MyPlanning Item"
                    },
                    description: "My description",
                    totalAmount: 100,
                    date: "01.01.1970"
                };
            }
        },
        isConnected: {
            type: Boolean,
            default: false
        },
        hasDate: {
            type: Boolean,
            default: true
        },
        clickEnabled: {
            type: Boolean,
            default: true
        },
        editMode: {
            type: Boolean,
            default: false
        },
        costCenterData: {
            type: Array,
            required: false,
            default() {
                return [];
            }
        }
    },
    data() {
        return {
            localEditMode: this.editMode,
            displayOverlay: false,
            localPlanningItem: this.planningItem,
            originalPlanningItem: _.clone(this.planningItem),
            date: moment(this.planningItem.date).toDate(),
            vCalendarAttributes: [
                {
                    key: "today",
                    dot: "red",
                    dates: new Date(),
                },
            ]
        };
    },
    validations: {
        localPlanningItem: {
            costCenter: {
                id: {
                    required,
                    integer,
                    notDefaultCostCenterId
                }
            },
            totalAmount: {
                required,
                notZero
            }
        },
    },
    computed: {
        negativePostiveClass() {
            return {
                "negative": this.localPlanningItem.totalAmount < 0,
                "positive": this.localPlanningItem.totalAmount >= 0
            };
        },
        startDate() {
            return moment(this.year + "-" + this.month + "-1").toDate();
        },
        endDate() {
            return moment(this.year + "-" + this.month + "-1").endOf("month").toDate();
        }
    },
    watch: {
        localEditMode () {
            //remember original value
            this.originalPlanningItem = _.clone(this.localPlanningItem);
        }
    },
    mounted() {
        this.$watch(
            "localPlanningItem.totalAmount", //what you want to watch
            () => {
                if (this.localPlanningItem.totalAmount === "0"){
                    this.localPlanningItem.totalAmount = "";
                }
            }
        );
    },
    methods: {
        errorClass(validator) {
            return {
                "is-invalid": validator.$error,
            };
        },
        showOverlay() {
            if (this.clickEnabled){
                this.displayOverlay = true;
            }
        },
        closeOverlay() {
            this.displayOverlay = false;
            this.$emit("close", this.localPlanningItem.id);
        },
        deletePlanning() {
            this.$emit("delete", this.localPlanningItem.id);
        },
        connectPlanning() {
            this.$emit("connect", this.localPlanningItem.id);
        },
        editPlanning() {
            this.localEditMode = true;
        },
        cancelEdit(){
            if (!this.localPlanningItem.isNew){
                this.localPlanningItem = this.originalPlanningItem;
                this.localEditMode = false;
                this.displayOverlay = false;
                this.$emit("cancel", this.localPlanningItem);
            }else{
                this.deletePlanning();
            }
        },
        saveEdit(){
            this.$v.$touch();
            if (!this.$v.$invalid) {
                this.localPlanningItem.date = this.date;
                //workaround to avoid multiple creation of plannign elements when the element gets created via the
                //ui and edit it again before freshing the page.
                this.localPlanningItem.isNew = this.planningItem.isNew;
                this.localEditMode = false;
                this.displayOverlay = false;
                this.$emit("save", this.localPlanningItem, true);
            }
        },
        costCenterUpdate(item){
            this.$v.localPlanningItem.costCenter.id.$touch();
            this.localPlanningItem.costCenter = item;
        },
        createCostCenter(newCostCenterName){
            this.$refs.topProgress.start();

            window.axios.post("/api/finance/costCenter", {title: newCostCenterName})
                .then(res => {
                    this.$emit("createCostCenter", res.data);
                    this.$refs.topProgress.done();
                });

        }
    }
};
</script>

<style scoped>
    .planningElement {
        cursor: pointer;
    }
    .date {
        font-size: 80%;
    }
    .planningElement {
        border: solid #adb5bd 1px;
        padding: 8px;
        margin: 0 0 1em 0;
    }

    .planningElement .description {
        font-size: 70%;
        margin: 0;
        word-wrap: break-word;
    }

    .positive {
        color: green;
    }

    .negative {
        color: red;
    }

    .title {
        position: relative;
    }

    .title .float-left {
        width: 70%
    }

    .title .float-right {
        width: 30%
    }

    .pre-formatted {
        white-space: pre-wrap;
    }
</style>
