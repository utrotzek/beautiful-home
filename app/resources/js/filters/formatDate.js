import moment from "moment";

//TODO: set format in moment
export default function formatDate (value) {
    if (value) {
        if (typeof value === "object"){
            return moment(value.date).format("DD.MM.YYYY");
        }else{
            return moment(String(value)).format("DD.MM.YYYY");
        }
    }
}
