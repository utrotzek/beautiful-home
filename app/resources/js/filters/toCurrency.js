export default function toCurrency (value) {
    if (typeof value !== "number") {
        return value;
    }
    var formatter = new Intl.NumberFormat("de-DE", {
        style: "currency",
        currency: "EUR",
        minimumFractionDigits: 2
    });
    return formatter.format(value);
}
