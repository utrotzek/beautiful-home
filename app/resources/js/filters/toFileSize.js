export default function toFileSize(value){
    if (isNaN(value)){
        return "undefined size";
    }else{
        const units = [ "byte", "kB", "MB", "GB", "TB"];
        let newValue = value;
        let counter = 0;

        if (newValue >= 1000){
            while (newValue >= 1000){
                newValue = newValue / 1024;
                counter++;
            }
        }

        var formatter = new Intl.NumberFormat("de-DE", {
            minimumFractionDigits: 1,
            maximumFractionDigits: 1
        });
        return formatter.format(newValue) + " " + units[counter];
    }
}
