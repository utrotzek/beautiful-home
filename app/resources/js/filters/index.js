import toCurrency from "./toCurrency";
import formatDate from "./formatDate";
import toFileSize from "./toFileSize";

export default {
    install(Vue) {
        Vue.filter("toCurrency", toCurrency);
        Vue.filter("formatDate", formatDate);
        Vue.filter("toFileSize", toFileSize);
    }
};
