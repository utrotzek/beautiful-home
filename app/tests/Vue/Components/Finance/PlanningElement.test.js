import "../Mocks/MatchMedia.mock";
import _ from "lodash";
import {mount} from "@vue/test-utils";
import PlanningElement from "../../../../resources/js/components/Finance/PlanningElement";
import Vue from "vue";
import Vuelidate from "vuelidate";
import filters from "../../../../resources/js/filters";
import VCalendar from "v-calendar";
import PlanningDataFixtures from "./Fixtures/PlanningFixtures";
import CostCenterFixtures from "./Fixtures/CostCenterFixtures";

Vue.use(filters);
Vue.use(Vuelidate);
Vue.use(VCalendar);

describe("PlanningElement", () => {
    let wrapper = null;
    let planningData = null;

    beforeEach(() => {
        planningData = _.clone(PlanningDataFixtures[0]);
        wrapper = mount(PlanningElement,{
            propsData: {
                year: 3000,
                month: 1,
                planningItem: planningData,
                isConnected: false,
                hasDate: true,
                editMode: false,
                costCenterData: CostCenterFixtures,
                allowConnection: true

            }
        });
    });

    it ("renders properly", () => {
        expect(wrapper.element).toMatchSnapshot();
    });

    it ("renders negative/positive class propery", () => {
        wrapper.setData(
            {
                localPlanningItem: {
                    totalAmount: -100
                }
            }
        );
        expect(wrapper.vm.negativePostiveClass).toEqual({negative: true, positive: false});

        wrapper.setData(
            {
                localPlanningItem: {
                    totalAmount: 100
                }
            }
        );
        expect(wrapper.vm.negativePostiveClass).toEqual({negative: false, positive: true});
    });

    it ("calculates start date properly", () => {
        const expectedDate = "1.1.3000";
        const date = wrapper.vm.startDate;
        const actualDate = date.getDate() + "." + (date.getMonth()+1) + "." + date.getFullYear();
        expect(actualDate).toEqual(expectedDate);
    });

    it ("calculates end date properly", () => {
        const expectedDate = "31.1.3000";
        const date = wrapper.vm.endDate;
        const actualDate = date.getDate() + "." + (date.getMonth()+1) + "." + date.getFullYear();
        expect(actualDate).toEqual(expectedDate);
    });

    it ("will set total amount to an empty value if the value is set to zero",async () =>{
        wrapper.vm.localPlanningItem.totalAmount = "0";
        await Vue.nextTick();
        expect(wrapper.vm.localPlanningItem.totalAmount).toEqual("");
    });
    //TODO: test validation

    it ("Can display and close the button overlay", async () => {
        //expect overlay is closed initially
        expect(wrapper.find(".overlay").exists()).toBeFalsy();

        //open overlay
        wrapper.find(".show-overlay-button").trigger("click");
        await Vue.nextTick();
        expect(wrapper.find(".overlay").exists()).toBeTruthy();

        //close overlay
        wrapper.find(".overlay .close").trigger("click");
        await Vue.nextTick();

        const expectedId = planningData.id;
        expect(wrapper.emitted().close[0]).toEqual([expectedId]);
        expect(wrapper.find(".overlay").exists()).toBeFalsy();
    });

    it ("Can fire delete event", async () => {
        wrapper.vm.displayOverlay = true;
        await Vue.nextTick();

        wrapper.find(".overlay .delete").trigger("click");
        expectDeletion();
    });

    it ("Can fire connect event", async () => {
        wrapper.vm.displayOverlay = true;
        await Vue.nextTick();

        wrapper.find(".overlay .connect").trigger("click");

        const expectedId = planningData.id;
        expect(wrapper.emitted().connect[0]).toEqual([expectedId]);
    });

    it ("Edit mode can be enabled", async () => {
        wrapper.vm.displayOverlay = true;
        await Vue.nextTick();

        wrapper.find(".overlay .edit").trigger("click");
        expect(wrapper.vm.localEditMode).toBeTruthy();
    });

    it ("deletes new item on cancellation", async () => {
        wrapper.vm.localPlanningItem.isNew = true;
        wrapper.vm.displayOverlay = true;
        wrapper.vm.localEditMode = true;
        await Vue.nextTick();

        wrapper.find(".overlay .cancel").trigger("click");
        await Vue.nextTick();
        expectDeletion();
    });

    it ("resets existing item to initial values on cancellation", async () => {
        //initialize data and store original values
        planningData.isNew = false;
        const originalPlanningData = _.clone(planningData);
        wrapper.vm.localEditMode = true;
        await Vue.nextTick();

        //manipulate data
        wrapper.vm.localPlanningItem.totalAmount = 3000;
        wrapper.vm.localPlanningItem.description = "My new descrpition";
        wrapper.vm.displayOverlay = true;
        await Vue.nextTick();

        //click on "cancel" and expect values have been restored to original values
        wrapper.find(".overlay .cancel").trigger("click");
        await Vue.nextTick();
        expect(wrapper.vm.localPlanningItem).toStrictEqual(originalPlanningData);
    });

    it ("Will save and fires event", async () => {
        wrapper.vm.localEditMode = true;
        wrapper.vm.displayOverlay = true;
        await Vue.nextTick();

        //click an "cancel" and expect values have been restored to original values
        wrapper.find(".overlay .save").trigger("click");
        await Vue.nextTick();

        expect(wrapper.vm.localEditMode).toBeFalsy();
        expect(wrapper.vm.displayOverlay).toBeFalsy();
        expect(wrapper.emitted().save[0]).toEqual([planningData]);
    });

    it ("can update cost centers", async() => {
        wrapper.vm.costCenterUpdate(CostCenterFixtures[2]);
        expect(wrapper.vm.localPlanningItem.costCenter).toEqual(CostCenterFixtures[2]);
    });

    function expectDeletion() {
        const expectedId = planningData.id;
        expect(wrapper.emitted().delete[0]).toEqual([expectedId]);
    }
});
