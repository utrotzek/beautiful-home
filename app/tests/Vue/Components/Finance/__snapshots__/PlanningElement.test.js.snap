// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlanningElement renders properly 1`] = `
<div
  class="planningElement form-group show-overlay-button"
>
  <!---->
   
  <div
    class="date"
  >
    <span>
      
    </span>
  </div>
   
  <div
    class="title cost-center"
  >
    <div
      class="float-left"
    >
      <span>
        Einkaufen
      </span>
    </div>
     
    <div
      class="float-right text-right negative"
    >
      <span>
        -â‚¬100.00
      </span>
    </div>
  </div>
   
  <div
    class="clearfix"
  />
   
  <div
    class="description"
  >
    <span
      class="pre-formatted"
    >
      Eintrag 1
    </span>
  </div>
   
  <!---->
</div>
`;
